daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala

```diff
@@ -59,19 +96,53 @@ object Speedy {
       var localContracts: Map[V.ContractId, (Ref.TypeConName, SValue)],
       // global contract discriminators, that are discriminators from contract created in previous transactions
       var globalDiscriminators: Set[crypto.Hash],
+      /* Used when enableLightweightStepTracing is true */
+      var steps: Int,
+      /* Used when enableInstrumentation is true */
+      var track: Instrumentation,
   ) {
 
+    /* kont manipulation... */
+
+    @inline def kontDepth(): Int = kontStack.size()
+
     @inline def pushKont(k: Kont): Unit = {
-      kontStack.add(k); ()
+      kontStack.add(k)
+      if (enableInstrumentation) {
+        track.countPushesKont += 1
+        if (kontDepth > track.maxDepthKont) track.maxDepthKont = kontDepth
+      }
     }
 
     @inline def popKont(): Kont = {
       kontStack.remove(kontStack.size - 1)
     }
 
+    /* env manipulation... */
+
+    @inline def envDepth(): Int = env.size()
+
     @inline def getEnv(i: Int): SValue = env.get(env.size - i)
-    @inline def popEnv(count: Int): Unit =
+
+    @inline def pushEnv(v: SValue): Unit = {
+      env.add(v)
+      if (enableInstrumentation) {
```

Comment:
Are we sure we don't have to pay for this even when `enableInstrumentation` is false? Would it make sense to subclass `Speedy` and override these methods? Or do we have to pay for additional dynamic dispatch then?